### PR TITLE
Test/edit attention unit tests

### PR DESCRIPTION
**Pull request by kaili-yang**
## Summary

A follow-up PR of https://github.com/Comfy-Org/ComfyUI_frontend/issues/11107.

## Changes

Add unit test to `editAttention.ts` 
- [x] `Extract pure functions to module level`: **Moved** `incrementWeight`, `findNearestEnclosure`, and `addWeightToParentheses` out of the `init()` closure and **promoted** them to module-level functions with `export` to allow for independent testing.
- [x] `Add unit tests for incrementWeight`: **Added** 6 tests covering edge cases such as normal increment/decrement, NaN input, negative weights, and floating-point precision.
- [x] `Add unit tests for findNearestEnclosure`: **Added** 7 tests covering edge cases including simple brackets, no brackets, cursor outside, nested brackets (inner/outer), empty strings, and missing closing brackets.
- [x] `Add unit tests for addWeightToParentheses`: **Added** 6 tests covering scenarios like adding a default 1.0 weight, retaining existing weights, no changes when brackets are absent, scientific notation weights, negative weights, and multi-word tokens.
- [x] `Mock app module`: **Used** `vi.mock('@/scripts/app')` to intercept side effects from `app.registerExtension`, **preventing** the triggering of ComfyUI extension registration logic during module import.

---

> [!NOTE]
> **Medium Risk**
> Adjusts token selection/weight-detection logic used during Ctrl/Cmd+Arrow editing, which could subtly change how prompts are rewritten in edge cases (nested parens, scientific notation, time-like text). Adds tests that should reduce regression risk but behavior changes still warrant verification in the UI.
> 
> **Overview**
> Adds a new `vitest` unit test suite for `editAttention` by mocking `app.registerExtension` side effects and validating `incrementWeight`, `findNearestEnclosure`, and `addWeightToParentheses` across common and edge cases.
> 
> Refactors those helpers to exported module-level functions and tightens parsing/selection behavior: `findNearestEnclosure` now handles the cursor being on an opening `(`, `addWeightToParentheses` improves trailing weight detection (supports scientific notation/negatives and avoids misclassifying time-like `12:30`), and the weight-rewrite regex now matches exponent forms.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit df20340b492be3a2d85b8359d2eef64aa0239e03. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-11301-Test-edit-attention-unit-tests-3446d73d365081f29e8dfceefc06f5d3) by [Unito](https://www.unito.io)

### CODE REVIEW

File: src/extensions/core/editAttention.test.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      end: 13
+    })
+  })
+
+  it('finds the outer enclosure when cursor is on outer content', () => {
+    expect(findNearestEnclosure('(outer (inner) end)', 2)).toEqual({
+      start: 1,
+      end: 18
+    })
+  })
```
Comment (christian-byrne) [inline src/extensions/core/editAttention.test.ts:67]: Do you want to parameterize these?

See: https://vitest.dev/api/test.html#test-for

File: src/extensions/core/editAttention.test.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  it('returns null for empty string', () => {
+    expect(findNearestEnclosure('', 0)).toBeNull()
+  })
+
+  it('returns null when opening paren has no matching closing paren', () => {
+    expect(findNearestEnclosure('(cat', 2)).toBeNull()
+  })
+})
+
+describe('addWeightToParentheses', () => {
```
Comment (christian-byrne) [inline src/extensions/core/editAttention.test.ts:78]: Potentially also good to use https://vitest.dev/api/test.html#test-for here

File: src/extensions/core/editAttention.ts [hunk 3]
```diff
@@ -1,6 +1,61 @@
...
+
+  openCount = 0
+  closeCount = 0
+
+  while (end < text.length) {
+    if (text[end] === ')' && openCount === closeCount) break
+    if (text[end] === '(') openCount++
+    if (text[end] === ')') closeCount++
+    end++
+  }
```
Comment (christian-byrne) [inline src/extensions/core/editAttention.ts:40]: nit (not for this PR): I know this is only in the diff because you are copy-pasting the code into outer scope, but someone in the future should simplify this and have it use pure shared functions for things like "find closing parentheses" and "find opening parentheses" => move them into src/base => add unit tests and permuatation tests for them. @coderabbitai can you make a followup Issue to do this?

File: src/extensions/core/editAttention.ts [hunk 4]
```diff
@@ -138,18 +130,15 @@ app.registerExtension({
...
         selectedText[0] !== '(' ||
         selectedText[selectedText.length - 1] !== ')'
       ) {
         selectedText = `(${selectedText})`
       }
 
-      // If the selection does not have a weight, add a weight of 1.0
       selectedText = addWeightToParentheses(selectedText)
 
-      // Increment the weight
```
Comment (christian-byrne) [inline src/extensions/core/editAttention.ts:152]: I think removing comments is generally good -- as comments cannot be maintained via tests and may over time lie about the code.

But if we want to remove comments, I think we just need to verify that all the same concepts/semantics being expressed in them are properly expressed in the code (which is possible, but needs to be verified).

For example, consider this code block:

```typescript
      // If the selection is not enclosed in parentheses, add them
      if (
        selectedText[0] !== '(' ||
        selectedText[selectedText.length - 1] !== ')'
      ) {
        selectedText = `(${selectedText})`
      }
```

Instead of just removing the comment and leaving the rest unchanged, we can change the code to have all the same expressiveness of the comment:

```typescript
      const isSelectionNotEnclosedInParentheses = 
          selectedText[0] !== '(' ||
          selectedText[selectedText.length - 1] !== ')'
      if (isSelectionNotEnclosedInParentheses) selectedText = `(${selectedText})`
```

This is just an example, it's not necessarily something that this code block needs.

**Review by christian-byrne — COMMENTED** [review]
Looking good!

File: src/extensions/core/editAttention.ts [hunk 5]
```diff
@@ -1,6 +1,61 @@
...
+
+  while (end < text.length) {
+    if (text[end] === ')' && openCount === closeCount) break
+    if (text[end] === '(') openCount++
+    if (text[end] === ')') closeCount++
+    end++
+  }
+  if (end === text.length) return null
+
+  return { start: start + 1, end: end }
```
[the remark this answers is not in the document]
Comment (kaili-yang) [reply src/extensions/core/editAttention.ts:46]: Fixed — added an upfront check: if the cursor is exactly on `(`, we skip the backward scan and start the forward scan from `cursorPos + 1` with `start = cursorPos`, so the function correctly returns the enclosure.

File: src/extensions/core/editAttention.ts [hunk 6]
```diff
@@ -1,6 +1,61 @@
...
+  const parenMatch = text.match(parenRegex)
+
+  const floatRegex = /:([+-]?(\d*\.)?\d+([eE][+-]?\d+)?)/
+  const floatMatch = text.match(floatRegex)
+
+  if (parenMatch && !floatMatch) {
+    return `(${parenMatch[1]}:1.0)`
+  } else {
+    return text
+  }
```
[the remark this answers is not in the document]
Comment (kaili-yang) [reply src/extensions/core/editAttention.ts]: Fixed — replaced the broad `floatRegex` match against the whole text with a trailing-anchor regex applied to `innerText` only, plus a negative lookbehind `(?<!\d)` to prevent digit-separated values like `12:30` from being misidentified as weights; also updated the replacement regex in `init()` to support scientific notation.

File: src/extensions/core/editAttention.test.ts (hunk 1 above)
Comment (kaili-yang) [reply src/extensions/core/editAttention.test.ts]: Good point. Used `test-for` here. It's a cleaner implementation.

File: src/extensions/core/editAttention.test.ts (hunk 2 above)
Comment (kaili-yang) [reply src/extensions/core/editAttention.test.ts:82]: Used `test-for` here.

File: src/extensions/core/editAttention.ts (hunk 3 above)
Comment (kaili-yang) [reply src/extensions/core/editAttention.ts]: Not addressed in this PR — agreed this is a worthwhile future refactor,  leaving it to the follow-up issue coderabbitai created.

File: src/extensions/core/editAttention.ts (hunk 4 above)
Comment (kaili-yang) [reply src/extensions/core/editAttention.ts:152]: Good point. Added your opiton to my `skills.md`, so the intent is now expressed in code rather than comment.

File: src/extensions/core/editAttention.ts [hunk 7]
```diff
@@ -1,6 +1,60 @@
...
+  return { start: start + 1, end: end }
+}
+
+export function addWeightToParentheses(text: string): string {
+  const parenMatch = text.match(/^\((.*)\)$/)
+  if (!parenMatch) return text
+  const innerText = parenMatch[1]
+  const hasTrailingWeight = /(?<!\d):[+-]?(?:\d*\.)?\d+(?:[eE][+-]?\d+)?$/.test(
+    innerText
+  )
```
[the remark this answers is not in the document]
Comment (christian-byrne) [reply src/extensions/core/editAttention.ts]: issue (blocking): confirmed this is a real bug. The `(?<!\d)` lookbehind rejects the match when inner text ends with a digit before the colon (e.g. `v2:1.5`, `sdxl1:0.8`). This is common for model/LoRA names.

Repro: `addWeightToParentheses('(v2:1.5)')` returns `'(v2:1.5:1.0)'` instead of `'(v2:1.5)'`.

Fix: replace `(?<!\d)` with a word-boundary or anchor the match to the last colon-number sequence. Something like `:([+-]?(?:\d*\.)?\d+(?:[eE][+-]?\d+)?)$` without the lookbehind should work - the trailing `$` anchor already prevents the `12:30` false positive since `30` would match but then the full inner text `time 12:30` ends at the match, which is correct (it IS a trailing weight pattern). The real distinction is: `12:30` in `(time 12:30)` should get `:1.0` added, but `v2:1.5` in `(v2:1.5)` should not. You need a heuristic - e.g. require the weight portion to contain a decimal point or be preceded by the last colon in the string.

**Review by dante01yoon — CHANGES_REQUESTED** [review]
issue: I agree with the existing unresolved blocker on `addWeightToParentheses`. On current head 466c1e666dc4b76e05bb486a342139eca993c7e5, the `(?<!\d)` lookbehind still rejects valid trailing weights when the token name ends in a digit. I verified the current regex returns false for `v2:1.5` and `sdxl1:0.8`, so `addWeightToParentheses("(v2:1.5)")` would append another default weight instead of preserving the existing one. That can regress common prompt, model, or LoRA names.

Could we adjust the trailing-weight heuristic and add focused coverage for digit-ending weighted tokens such as `(v2:1.5)` and `(sdxl1:0.8)`, while keeping the intended `(time 12:30)` behavior?

Validation: GitHub CI is green, and `pnpm exec vitest run src/extensions/core/editAttention.test.ts` passed 21 tests in a clean worktree. The focused suite passes because it does not cover this digit-ending token branch yet.

**Review by kaili-yang — COMMENTED** [review]
Good catch. The `(?<!\d)` lookbehind was too broad — it blocked any colon preceded by a digit, which incorrectly rejected common model/LoRA names like `v2:1.5` and `sdxl1:0.8`.

**Fix:** replaced the lookbehind with an explicit time-pattern guard:
```ts
const looksLikeTime = /(?:^|\s)\d{1,2}:\d{2}$/.test(innerText)
const hasTrailingWeight = !looksLikeTime && /:[+-]?(?:\d*\.)?\d+(?:[eE][+-]?\d+)?$/.test(innerText)
```

The key insight: a time like `12:30` is always preceded by whitespace (or is at the start of the string), whereas a token-weight colon in `v2:1.5` or `sdxl1:0.8` follows non-whitespace characters directly.

Added two new test cases for the digit-ending names you mentioned:
- `(v2:1.5)` → unchanged ✓
- `(sdxl1:0.8)` → unchanged ✓

All 23 tests pass.

**Review by dante01yoon — APPROVED** [review]
Approved current head 74a8c0ef50aa86c46df71491a8fc60395e743c3f. The previous digit-ending token blocker is fixed: the lookbehind was replaced with a time-like guard, and the focused tests now cover (v2:1.5) and (sdxl1:0.8) while preserving the time-string behavior. Validation: GitHub checks are green, and pnpm exec vitest run src/extensions/core/editAttention.test.ts passed 23/23 in a temporary PR worktree.

**Comment by dante01yoon** [comment]
@coderabbitai re-review and approve if no issue

**Comment by dante01yoon** [comment]
@coderabbitai approve
